src/test/java/org/terasology/launcher/repositories/JenkinsClientTest.java

```diff
@@ -0,0 +1,53 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.launcher.repositories;
+
+import com.google.gson.Gson;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Note: We are testing with the default Gson parser here. This can go out of sync with the instantiation of
+ * JenkinsClient in RepositoryManager. Using common dependency injection for productive code and tests would ensure we
+ * are using the "real thing" under test here.
+ */
+@DisplayName("JenkinsClient")
+class JenkinsClientTest {
+
+    @Test
+    @DisplayName("should handle IOException on request(url) gracefully")
+    void nullOnIoException() throws IOException {
+        final Gson gson = new Gson();
+        final JenkinsClient client = new JenkinsClient(gson);
+
+        URL urlThrowingException = mock(URL.class);
+        doThrow(IOException.class).when(urlThrowingException).openStream();
+
+        assertNull(client.request(urlThrowingException));
+    }
+
+    @Test
+    @DisplayName("can handle invalid JSON payload")
+    void canHandleInvalidJsonPayload() throws IOException {
+        final Gson gson = new Gson();
+        final JenkinsClient client = new JenkinsClient(gson);
+
+        InputStream invalidPayload = new ByteArrayInputStream("{ this is ] no json |[!".getBytes());
```

Comment:
I'm not sure, maybe it would be better to stub the resoponse of `gson.fromJson(...)` here to force throwing an exception... 🤔

Reply:
As long as it's a few objects that are easy to create (e.g. not something like MTE with a ton of start-up time and interconnected state), it is nice to be able to use real objects instead of mockito mocks.

Partly for the reason you mentioned: it can be hard to know when your mockito-defined behavior is no longer realistic.

and partly for some Java-specific worries I caught when working on old Mockito tests with warnings about illegal reflective access. Maybe they've fixed that? because Mockito is clearly still going strong and hasn't been destroyed by new releases of Java.